ziprofs.py

```diff
@@ -162,19 +157,44 @@ def getattr(self, path, fh=None):
 
     def open(self, path, flags):
         zip_path = self.get_zip_path(path)
-        if not zip_path:
-            return os.open(path, flags)
-        return 0
+        if zip_path:
+            with self._lock:
+                fh = self._get_free_zip_fh()
+                zf = self.zip_factory.get(zip_path)
+                self._zip_file_fh[fh] = zf.open(path[len(zip_path) + 1:])
+                return fh
+        else:
+            return os.open(path, flags) << 1
 
     def read(self, path, size, offset, fh):
-        with self.rwlock:
-            zip_path = self.get_zip_path(path)
-            if not zip_path:
-                os.lseek(fh, offset, 0)
-                return os.read(fh, size)
-            zf = self.zip_factory.get(zip_path)
-            subpath = path[len(zip_path)+1:]
-            return zf.read(subpath, size, offset)
+        if self.get_zip_path(path):
+            f = self._zip_file_fh[fh]  # should be here (file is first opened, then read)
+            with f.ziplock():
+                if f.seekable():
+                    if self.log.isEnabledFor(logging.DEBUG):
+                        foffset = f.tell()
+                        self.log.debug(f" file offset: {foffset}, read offset: {offset}, diff: {offset-foffset}")
+                    f.seek(offset)
+                    return f.read(size)
+                else:
+                    # emulate seek by reading and discarding chunks
+                    # zipfile.py#1110 - tell()
+                    # TODO: remove usage of hidden fields
+                    filepos = f._orig_file_size - f._left - len(f._readbuffer) + f._offset
+                    offset -= filepos
+                    if offset < 0:
+                        # cant seek back in non-seekable file
+                        raise FuseOSError(errno.EINVAL)
```

Comment:
EIO seems more appropriate but I'm not sure.

Reply:
I would leave it.
https://manpage.me/index.cgi?apropos=0&q=read&sektion=2&manpath=Debian+8.1.0&arch=default&format=html
ERRORS section
IMO It seems to be most valid